src/client/pfs.go

```diff
@@ -505,6 +501,15 @@ func (c APIClient) SubscribeCommitF(repo, branch, from string, state pfs.CommitS
 	}
 }
 
+// PutObjectAsync puts a value into the object store asynchronously.
+func (c APIClient) PutObjectAsync(tags []*pfs.Tag) (*PutObjectWriteCloserAsync, error) {
+	w, err := c.newPutObjectWriteCloserAsync(tags)
+	if err != nil {
+		return nil, grpcutil.ScrubGRPC(err)
+	}
+	return w, nil
+}
```

Comment:
Having async methods in Go is a bit of a weird smell to me. Since the caller can always make things async by wrapping it in a `go` directive, and it seems to promote dropping errors. Also looking at the code it looks like this is indeed `async` but the more significant feature of the async writer is that it's buffered.

Reply:
Actually reading the code below I think we should just call this `PutObjectWriteCloser`, it's only `async` in that `Write` calls return immediately before anything is actually written, but that's an assumed feature of `WriteCloser`s, `os.File` can behave the same way. That might give us a naming conflict in which case, maybe we call this `BufferedPutObjectWriteCloser`, or just eliminate the other implementation?

Reply:
So I think we discussed this before, but I think there is actually a pretty good argument for having asynchronous writers in Go particularly because go routines are so cheap, it is much cleaner and not really that error prone considering it is a reasonable expectation for a user of the writer to close the writer when they are done. I believe when we agreed on this pattern, it was mainly because this is how the gcs client implements its writers. There is a go routine that is reading from a pipe that handles the writes. I think this code is bound to get changed/scrapped anyways when we pull out the sidecar, so I am not too inclined to spend too much time on it if there are not any bugs since it will probably be changed again.

Reply:
I'm all for async writers like this. I don't disagree on anything except naming. I don't think you need `async` in the name of this type, `WriteCloser`s are assumed to have async semantics like this by golang convention.